docs/source/miscellaneous/doc/adr/0016-use-tickets-encoding-for-transactions.md

```diff
@@ -499,7 +499,193 @@ the worker pool has a finite size and even with the full batching algorithm in t
+target schema version to a specific version (which at time of writing can be 1 or 2), a transform will be run
+setting the transactions schema version going forward to that version.
+
+This is optional; if not configured, we will not attempt to run any transforms so the transaction schema version will
+remain what it is. If this was never configured at all, we default to using transactions1, to be consistent with
+previous behaviour.
+
+### Live Migrations
+
+Transactions2 was written with some of the heaviest users of AtlasDB in mind. These are core Palantir services where
+shutdown upgrades are costly or even verboten, and we thus implemented a mechanism for performing upgrades without
+downtime. The key problem here involves reasoning about service nodes running different binary versions that
+may be running different AtlasDB versions, that are concurrently operating on an AtlasDB deployment.
+
+#### Legacy Versions
+
+AtlasDB is deployed as a library in client services. Versions of AtlasDB may be partitioned based on their support
+for Transactions2 into the following categories:
+
+- **(A1) I always use transactions1.** These versions don't even know that the coordination service exists, and they
+  will read and write from transactions1 regardless of what other nodes are doing.
+- **(A2) I know about the coordination service, but not about transactions2.** These versions of AtlasDB will check
+  the coordination service, but they don't know how to read from/write to transactions2, and so they will throw if
+  they need to write to transactions2, or if they need to read the commit timestamp of a value written to
+  transactions2.
+- **(A3) I know about the coordination service and about transactions2.** This is fairly self-explanatory.
+
+Client services come with a version of AtlasDB, and at some point they typically enable transactions2 by default.
+We can thus similarly partition them:
+
+- C1 and C2 are defined as client service versions using versions of AtlasDB that would be classified as A1 and A2
+  respectively.
+- C3 refers to client service versions that use an A3 version of AtlasDB, but don't enable transactions2.
+- C4 refers to client service versions that use an A3 version of AtlasDB and enable transactions2. (Transactions2
+  cannot be enabled with A1 or A2 versions of AtlasDB, as it doesn't exist in those versions!)
+
+Now, we can consider how multiple client service versions interact:
+
+- Any sets of versions that don't include C4s are safely interoperable, because in these versions transactions2 is never
+  installed anyway, so we always use transactions1. Even C1s and C3s can safely interoperate, since although C1s never
+  read the coordination table, their effective assumption that it contains everything mapping to 1 is correct.
+- C1s and C4s running concurrently can lead to **SEVERE DATA CORRUPTION**, since the C4s might install transactions2
```

Comment:
Having C4 at and point in the past makes it potentially unsafe to roll back to C1 even if not running them concurrently.

Reply:
Added mention of how _if_ transactions2 has been used at some point in the past _then_ C1s can result in **SEVERE DATA CORRUPTION**™️

Reply:
There is a strange technical argument where if you ran with a C4 but it never did a transaction... 😅